pkg/capabilities/consensus/ocr3/reporting_plugin.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return nil, err
+	}
+
+	rids := []string{}
+	for _, q := range queryReq.RequestIds {
+		rids = append(rids, q.RequestId)
+	}
+
+	reqs, err := r.s.getN(ctx, rids)
+	if err != nil {
+		return nil, err
+	}
+
+	obs := &pbtypes.Observation{}
+	for _, r := range reqs {
+		obsPb, err := r.Observations.Proto()
+		if err != nil {
+			// TODO: we can't process this observation: log and continue
+		}
+
+		r := &pbtypes.Request{
+			Observation: obsPb,
+			RequestId: &pbtypes.RequestId{
+				RequestId:   r.RequestID,
+				NamespaceId: r.NamespaceID,
+			},
+		}
+
+		obs.Requests = append(obs.Requests, r)
+	}
+
+	return proto.Marshal(obs)
+}
+
+func (r *reportingPlugin) ValidateObservation(outctx ocr3types.OutcomeContext, query types.Query, ao types.AttributedObservation) error {
+	// TODO : just try and unmarshal the message
+	// basic validation
+	return nil
+}
+
+func (r *reportingPlugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.Query) (ocr3types.Quorum, error) {
+	return ocr3types.QuorumTwoFPlusOne, nil
+}
+
+func (r *reportingPlugin) Outcome(outctx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	m := map[string]map[ocrcommon.OracleID][]values.Value{}
+	for _, o := range aos {
+		obs := &pbtypes.Observation{}
+		err := proto.Unmarshal(o.Observation, obs)
+		if err != nil {
+			// TODO log
+			continue
+		}
+
+		for _, r := range obs.Requests {
+			rid := r.RequestId.RequestId
+			if _, ok := m[rid]; !ok {
+				m[rid] = make(map[ocrcommon.OracleID][]values.Value)
+			}
+
+			val, err := values.FromProto(r.Observation)
+			if err != nil {
+				continue
+			}
+
+			m[rid][o.Observer] = append(m[rid][o.Observer], val)
+		}
+	}
+
+	q := &pbtypes.Query{}
+	err := proto.Unmarshal(query, q)
+	if err != nil {
+		return nil, err
+	}
+
+	o := &pbtypes.Outcome{}
+	err = proto.Unmarshal(outctx.PreviousOutcome, o)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rid := range q.RequestIds {
+		requestObs, ok := m[rid.RequestId]
+		if !ok {
+			continue
+		}
+
+		namespaceOutcome, ok := o.Outcomes[rid.NamespaceId]
+		if !ok {
+			continue
+		}
+
+		outcome, err := r.aggregator.Aggregate(namespaceOutcome, requestObs)
+		if err != nil {
+
+		}
+
+		o.Outcomes[rid.NamespaceId] = outcome
+	}
+
+	o.CurrentQuery = q
+	return proto.Marshal(o)
+}
+
+func (r *reportingPlugin) Reports(seqNr uint64, outcome ocr3types.Outcome) ([]ocr3types.ReportWithInfo[any], error) {
+	o := &pbtypes.Outcome{}
+	err := proto.Unmarshal(outcome, o)
+	if err != nil {
+		return nil, err
+	}
+
+	reports := []ocr3types.ReportWithInfo[any]{}
+	// This doesn't handle a query which contains the same namespaceId multiple times.
+	for _, rid := range o.CurrentQuery.RequestIds {
+		o, ok := o.Outcomes[rid.NamespaceId]
+		if !ok {
+			continue
+		}
+
+		if !o.ShouldReport {
+			continue
+		}
+
+		mv, err := values.FromMapValueProto(o.EncodableOutcome.GetMapValue())
```

Comment:
I realized that this is incorrect. The fact that an ID was in query and has ShouldReport set doesn't mean it was even processed! It's possible that we didn't have enough observations or aggregate errored out. I think it's safer to just attach a list of IDs where the call to Aggregate returned ShouldReport=true in the latest round, instead of the query.